lib/packetgen/header/http/request.rb

```diff
@@ -0,0 +1,92 @@
+      # * the version ({Types::String}).
+      # * associated http headers ({Types::String}).
+      #
+      # == Create a HTTP Request header
+      #   # standalone
+      #   http_rqst = PacketGen::Header::HTTP::Request.new
+      #   # in a packet
+      #   pkt = PacketGen.gen("IP").add("TCP").add("HTTP::Request")
+      #   # access to HTTP Request header
+      #   pkt.http_request # => PacketGen::Header::HTTP::Request
+      #
+      # == HTTP Request attributes
+      #	  http_rqst.version = "HTTP/1.1"
+      #	  http_rqst.method = "GET"
+      #	  http_rqst.path = "/meow.html"
+      #   http_rqst.headers = "Host: tcpdump.org"     # string or
+      #	  http_rqst.headers = { host: "tcpdump.org" } # even a hash
+      #
+      # @author Kent 'picat' Gruber
+      class Request < Base
+        # @!attribute method
+        #   @return [Types::String]
+        define_field :method,  Types::String
+        # @!attribute path 
+        #   @return [Types::String]
+        define_field :path,    Types::String
+        # @!attribute version 
+        #   @return [Types::String]
+        define_field :version, Types::String, default: "HTTP/1.1"
+        # @!attribute headers
+        #   associated http/1.1 headers
+        #   @return [Types::HTTPHeaders]
+        define_field :headers, Types::HTTPHeaders 
+
+        # @param [Hash] options
+        # @option options [String] :method
+        # @option options [String] :path
+        # @option options [String] :version
+        # @option options [Hash]   :headers
+        def initialize(options={})
+          super(options)
+          self.headers ||= options[:headers]
+        end
+
+        # Give it a pretty name, the only way I know how.
+        # @return [String]
+        def protocol_name
+          "HTTP_Request"
+        end
+
+        # Read in the HTTP portion of the packet, and parse it. 
+        # @return [PacketGen::HTTP::Request]
+        def read(str)
+          # prepare data to parse
+          str = str.split("\n").map(&:strip).reject(&:empty?)
+          first_line = str.shift.split
+          self[:method]  = first_line[0]
+          self[:path]    = first_line[1]
+          self[:version] = first_line[2]
+          headers = str.join("\n")
+          self[:headers].read(headers)
+          self
+        end
+
+        # String representation of data.
+        # @return [String]
+        def to_s
+          raise FormatError, "Missing #method."  if self.method.empty?
+          raise FormatError, "Missing #path."    if self.path.empty?
+          raise FormatError, "Missing #version." if self.version.empty?
+          str = "" # build 'dat string
+          str << self[:method] << " " << self[:path] << " " << self[:version] << "\r\n" << self[:headers].to_s
+        end
+      end
+    end
+
+    self.add_class HTTP::Request 
+    TCP.bind_header HTTP::Request, body: ->(b) { /^(CONNECT|DELETE|GET|HEAD|OPTIONS|PATCH|POST|PUT)/ =~ b }
+  end
```

Comment:
This will not automatically set TCP port to 80 on packet generation. You may do:

```ruby
# HTTP::Request if dport == 80 and body begins with a HTTP verb
TCP.bind_header HTTP::Request, op: :and, dport: 80, body: ->(b) { /^(CONNECT|DELETE|GET|HEAD|OPTIONS|PATCH|POST|PUT)/ =~ b }
# HTTP::Request if sport == 80 and body begins with a HTTP verb
TCP.bind_header HTTP::Request, op: :and, sport: 80, body: ->(b) { /^(CONNECT|DELETE|GET|HEAD|OPTIONS|PATCH|POST|PUT)/ =~ b }
```